### DISCUSSION TITLE
`README.md` guide to set up coc.nvim completely in .lua

### DISCUSSION THREAD

**Discussion by tapyu** [neoclide/coc.nvim#3776] [General]
I am trying to start neovim with a setup totally in `.lua` and I have notice that `coc.nvim` doesn't have a guide (for `packer.nvim`, at least) to install and configure coc with `lua`, am I wrong?  Is `coc.nvim` not interested in doing this? As a beginner, I do not have ability to make out how to set this up in `lua` without a guide. Thank you in advance.

**Comment by chemzqm** [comment]
I don't have interest with configure nvim by lua, but PR welcome.

**Reply by gi1242** [reply]
To set up `coc.nvim` in lua just put the following in `lua/plugins.lua`

    return require('packer').startup( function(use)
        -- Plugins here.
        use { 'neoclide/coc.nvim', branch='release' }
        use 'lervag/vimtex'

        --[[ On new systems install COC packages via
            CocInstall coc-vimtex coc-texlab
        --]]
    end )
    
After editing the file do:

    source % | PackerClean | PackerInstall | PackerCompile

Finally, in your `init.vim` put

    lua require('plugins')

**Reply by wangmeng1125** [reply]
yes,man your way is so good , I tried it works very well

**Reply by akshaynaresh** [reply]
thanks it worked

**Comment by sigmaSd** [comment]
Someone seem to have ported the coc configuration to lua https://github.com/samhvw8/mvim/blob/master/lua/core/coc.lua I tried it and it seems to just work

**Comment by findNextStep** [comment]
if someone seach for "more lua" config about `<tab>` and `<CR>` here is my work

see https://github.com/findNextStep/mydot/blob/master/nvim/lua/plugin/coc.lua
```lua
local function check_back_space()
    local col = vim.fn.col('.') - 1
    return col == 0 or vim.fn.getline('.'):sub(col, col):match('%s') ~= nil
end

local opts = { silent = true, noremap = true, expr = true, replace_keycodes = false }

vim.keymap.set("i", "<Tab>",
    function()
        if vim.fn['coc#pum#visible']() == 1 then
            return vim.fn['coc#pum#next'](1)
        end
        if check_back_space() then
            return vim.fn['coc#refresh']()
        end
        return "<Tab>"
    end
    , opts)
vim.keymap.set("i", "<S-Tab>", function()
        if vim.fn['coc#pum#visible']() == 1 then
            return vim.fn['coc#pum#prev'](1)
        end
        return "<S-Tab>"
end, opts)
vim.keymap.set("i", "<CR>", function()
        if vim.fn['coc#pum#visible']() == 1 then
            return vim.fn['coc#pum#confirm']();
        end
       return "\r"
end, opts)
```

and if you want add coc_status into stateline, check [here](https://github.com/findNextStep/mydot/blob/master/nvim/lua/plugin/coc.lua#L174). we need `string.gsub(vim.g.coc_status, "%%", "%%%%")` for [format cause](https://zhuanlan.zhihu.com/p/600251060)

**Reply by akshaynaresh** [reply]
Thanks it worked

**Reply by PanchoBubble** [reply]
Nice one dude thanks a lot

**Comment by NeoSahadeo** [comment]
For anyone in the future. This is my config for Enter select.

> You need to specify `expr = true` for options
```lua
local coc_opts = {
    silent = true,
    noremap = true,
    expr = true,
}
```
The keybind
```lua
-- COC keybinds
vim.keymap.set('i', '<Cr>',
function() 
    -- Returns 1 if visible, 0 if not
    is_visible = vim.fn['coc#pum#visible']()
    if is_visible == 1 then
        return vim.fn['coc#pum#confirm']()
    else
        return '<Cr>'
    end
end, coc_opts)
```
